server/routes/webhooks/stripe-charge-failed.js

```diff
@@ -0,0 +1,106 @@
+const boom = require('boom');
+const stripe = require('../stripe');
+const basket = require('../../lib/basket-queue');
+
+// values that never change between webhook requests
+const signatureHeader = 'stripe-signature';
+const event_type = 'charge.failed';
+const frequency = 'monthly';
+const recurring = true;
+const service = 'stripe';
+
+// Error/Success message constants
+const forbiddenError = 'An error occurred while verifying the webhook signing secret';
+const incorrectEventError = 'This hook only processes charge failed events';
+const expandChargeError = 'An error occurred while expanding the charge';
+const retrieveSubscriptionErrorMsg = 'An error occurred while fetching the subscription';
+const notRecurringCharge = 'This hook only processes recurring charges that fail';
+const notPartOfSubscription = 'This charge is not part of a subscription';
+const successMessage = 'charge failed event processed';
+
+let stripeChargeFailed = (request, reply) => {
+  // Validate the Webhook signature
+  let endpointSecret = process.env.STRIPE_WEBHOOK_SIGNATURE_CHARGE_FAILED;
+  let signature = request.headers[signatureHeader];
+
+  let event = stripe.constructEvent(request.payload, signature, endpointSecret);
+
+  // If there's no event object, the signature verification failed
+  if (!event) {
+    return reply(boom.forbidden(forbiddenError));
+  }
+
+  // Only process charge.failed webhook events
+  if (event.type !== event_type) {
+    return reply(incorrectEventError);
+  }
+
+  let {charge} = event.data.object;
+
+  // If the charge has no invoice ID, this is not a recurring charge
+  if (!charge.invoice) {
+    return reply(notRecurringCharge);
+  }
+
+  // Expand the invoice object on the charge
+  stripe.retrieveCharge(charge.id, (fetchChargeError, charge) => {
+    if (fetchChargeError) {
+      return reply(boom.badImplementation(`${expandChargeError}: ${fetchChargeError}`));
+    }
+
+    // Unlikely, but ensure there's a subscription object to work with
+    if (!charge.invoice || !charge.invoice.subscription) {
+      return reply(notPartOfSubscription);
+    }
+
+    let {customer, subscription} = charge.invoice;
+
+    // Expand the subscription object
+    stripe.retrieveSubscription(customer, subscription, (retrieveSubscriptionError, subscription) => {
+      if (retrieveSubscriptionError) {
+        return reply(boom.badImplementation(`${retrieveSubscriptionErrorMsg}: ${retrieveSubscriptionError}`));
+      }
+
+      // Capture the variables we need to pass to the CRM
+      // Since this charge failed, we need to pass enough information
+      // for the CRM to create a new Opportunity record, and mark it
+      // immediately as 'Lost'
+      let {
+        currency,
+        created,
+        id: transaction_id,
```

Comment:
why destructure into `id` instead of leaving it as `transaction_id`? It doesn't seem like you use `id` anywhere but you redeclare it below and reassign it to `subscription_id`

Reply:
You've got your understanding of how this destructuring assignment works:

This line of code will look for `id` on the right operand (`charge`), and assign it's value to `transaction_id`. This is understandably confusing, especially if you try to liken this syntax to that of Object literals.

Here's a sample I ran in my browse console to illustrate:
![](https://i.imgur.com/DkgKbzU.png)

Reply:
heh, yep, you're right, I totally read the docs wrong